runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -409,6 +428,46 @@ function LanguageTree:_add_injections()
   return query_time
 end
 
+--- Run an asynchronous parse, calling {on_parse} when complete.
+---
+--- @param range boolean|Range|nil
+--- @param on_parse fun(trees: TSTrees)
+function LanguageTree:_async_parse(range, on_parse)
+  coroutine.resume(
+    --- @param co_range boolean|Range|nil
+    --- @param co_cb fun(trees: TSTrees)
+    coroutine.create(function(co_range, co_cb)
+      local co = coroutine.running()
+      local ct = self:_buf().changedtick
+      local unfinished = true
+      --- @type TSTrees
+      local trees
+
+      while unfinished do
```

Comment:
We should add some loop protection to avoid cases that hang or take too long to parse. For now we can just have a hard coded timeout limit which we may expose as an option in the future.

Reply:
could we use 'redrawtime' for that?

Reply:
No, I'd say this is different. `redrawtime` is a timeout ("stop doing anything after x ms"), while this is chunking ("don't block for more than x ms"). In particular, if we get things right, I don't expect anybody (reasonable) to have to adjust this.

Reply:
(We might _repurpose_ `redrawtime` after we have made ALL THE THINGS async, but that'll be a) a long way off and b) probably not bwc.)

Reply:
I could be wrong but I think here `redrawtime` *would* mean a timeout; the chunking would be the segment parse time, no?

Reply:
Oops, sorry, I misread. Yes, a hard "bailout" timeout as in Lewis' comment would make sense to tie to `redrawtime`.

Reply:
Question: should the callback still be fired if `redrawtime` is reached before parsing finishes? If so, `trees` will just be an empty list. For whatever reason, in the highlighter callback if we fire it even when `redrawtime` timeout is reached, running `for_each_tree` causes async parsing to continue, which means `redrawtime` is not respected. Of course we can just run a conditional check to only run the highlighter callback if `#trees ~= 0` but it still seems odd that this happens...

Reply:
Yes, the callback as a continuation which should always be called. It may be wise to pass back an error argument too which we can set to the timeout error/code.

Reply:
I don't think we have fully solved the "error" scenario for this, which this (unlike the async parsing) is part of. (There were some _very_ heated discussions in the past...)

Reply:
I opted to pass a boolean indicating whether timeout was reached; in treesitter.c if there is a timeout set, then parse() cannot throw a lua error so there are no other codes that could be needed right now

Reply:
Sorry, I know there was a discussion above about using `'redrawtime'`. I wasn't sure if my comment would be noticed.

I think that `'redrawtime'` is not the right option for limiting the total time of the parse.

The total time an async parse takes doesn't, in a major way, influence how responsive the editor is: the parse blocks for only 3ms and can execute during idle periods. Parsing a file even for 2 seconds is barely noticeable (ignoring the final parse with injections and the bug), so the limit should probably be much higher.

All other features that respect `'redrawtime'` can block the editor for an unknown amount of time, (and are not interruptible in any other way?).

Reply:
> Parsing a file even for 2 seconds is barely noticeable

Fair point, but nevertheless if a parse takes >2 seconds this is not something we should causally accept. So for now, 'redrawtime' is a reasonable limit.